File: src/parser_utils/record_utils.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

let loc_and_string_of_property_key key =
  let open Flow_ast in
  let open Expression.Object in
  match key with
  | Property.Computed _
  | Property.PrivateName _ ->
    failwith "records do not parse computed properties or private elements"
  | Property.Identifier (loc, { Identifier.name; _ }) -> (loc, name)
  | Property.StringLiteral (loc, { StringLiteral.value; _ }) -> (loc, value)
  | Property.NumberLiteral (loc, { NumberLiteral.value; _ }) ->
    (loc, Dtoa.ecma_string_of_float value)
  | Property.BigIntLiteral (loc, bigint) -> (loc, Flow_ast_utils.string_of_bigint bigint)

(* The set of record properties that have a default value supplied, e.g. `foo: number = 0,` *)
let defaulted_props_of_record record =
  let open Flow_ast.Statement.RecordDeclaration in
  let { body = (_, { Body.body; _ }); _ } = record in
  List.fold_left
    (fun acc -> function
      | Body.Property (_, { Property.key; default_value = Some _; _ }) ->
        let (_, name) = loc_and_string_of_property_key key in
        SSet.add name acc
      | _ -> acc)
    SSet.empty
    body

File: rust_port/crates/flow_parser_utils/src/record_utils.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeSet;

use dupe::Dupe;
use flow_common::js_number;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_parser::ast;
use flow_parser::ast_utils;

pub fn loc_and_string_of_property_key<Loc: Dupe>(
    key: &ast::expression::object::Key<Loc, Loc>,
) -> Option<(Loc, FlowSmolStr)> {
    use ast::expression::object::Key;
    match key {
        Key::Computed(_) | Key::PrivateName(_) => None,
        Key::Identifier(id) => Some((id.loc.dupe(), id.name.dupe())),
        Key::StringLiteral((loc, lit)) => Some((loc.dupe(), lit.value.dupe())),
        Key::NumberLiteral((loc, lit)) => Some((
            loc.dupe(),
            FlowSmolStr::new(js_number::ecma_string_of_float(lit.value)),
        )),
        Key::BigIntLiteral((loc, lit)) => Some((
            loc.dupe(),
            FlowSmolStr::new(ast_utils::string_of_bigint(lit)),
        )),
    }
}

pub fn defaulted_props_of_record<Loc: Dupe>(
    record: &ast::statement::RecordDeclaration<Loc, Loc>,
) -> BTreeSet<FlowSmolStr> {
    use ast::statement::record_declaration;

    record
        .body
        .body
        .iter()
        .filter_map(|element| {
            if let record_declaration::BodyElement::Property(record_declaration::Property {
                key,
                default_value: Some(_),
                ..
            }) = element
            {
                return loc_and_string_of_property_key(key).map(|(_, name)| name);
            }
            None
        })
        .collect()
}
